odo/backends/avro.py

```diff
@@ -0,0 +1,412 @@
+AVRO_TYPE_MAP.update(NAMED_TYPES_MAP)
+AVRO_TYPE_MAP.update(COMPOUND_TYPES_MAP)
+
+dshape_to_avro_primitive_types = {
+    ct.int8: 'bytes',
+    ct.int16: 'int',
+    ct.int32: 'int',
+    ct.int64: 'long',
+    ct.float32: 'float',
+    ct.float64: 'double',
+    ct.date_: 'long',
+    ct.datetime_: 'long',
+    ct.string: 'string',
+    ct.bool_: 'boolean'
+}
+
+
+class AVRO(object):
+    """Wrapper object for reading and writing an Avro container file
+
+    Parameters
+    ----------
+
+    uri : str
+        uri of avro data
+
+    schema : avro.schema.Schema
+        User specified Avro schema object.  Used to decode file or serialize new records to file.
+         schema is required to create a new Avro file.
+         If reading or appending to an existing Avro file, the writers_schema embedded in that file
+         will be used.
+
+    codec : str
+        compression codec.  Valid values: 'null', 'deflate', 'snappy'
+
+    """
+    def __init__(self, path, schema=None, codec='null', **kwargs):
+        self._path = path
+        self._schema = schema
+        self._codec = codec
+
+    def __iter__(self):
+        return self.reader.__iter__()
+
+    def next(self):
+        return self.reader.next()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        # Perform a close if there's no exception
+        if type is None:
+            self.reader.close()
+            self.writer.close()
+
+    def _get_writers_schema(self):
+        """
+        Extract writers schema embedded in an existing Avro file.
+        """
+        reader = self.reader
+        return schema_from_dict(self.reader.schema) if reader else None
+
+    path = property(lambda self: self._path)
+    codec = property(lambda self: self._codec)
+
+    @property
+    def schema(self):
+        if not self._schema:
+            sch = self._get_writers_schema()
+            if sch is None:
+                raise AvroException("Couldn't extract writers schema from '{0}'.  User must provide a valid schema".format(self.path))
+            self._schema = sch
+        return self._schema
+
+    @property
+    def reader(self):
+        if hasattr(self, '_reader'):
+            if hasattr(self, '_writer'):
+                self.flush()
+            return self._reader
+        else:
+            try:
+                reader_schema = self._schema.to_json() if self._schema else None
+                df_reader = fastavro.reader(
+                    open(self.path, 'rb'),
+                    reader_schema=reader_schema
+                )
+
+                return df_reader
+            except IOError as exc:
+                #If file doesn't exist, don't set _reader now.
+                #Allow for reevaluation later after file has been created.
+                #Otherwise, reraise exception
+                if exc.errno != errno.ENOENT:
+                    raise exc
+                return None
+
+    @staticmethod
```

Comment:
Is there a specific reason for the staticmethod here and below, rather than private top-level helper functions?  IMO we should convert these to module-scoped functions instead.